File: ruby/lib/mable_bank/infrastructure/csv_transfer_instruction_reader.rb
# frozen_string_literal: true

require 'csv'

module MableBank
  module Infrastructure
    class CsvTransferInstructionReader
      class ParseError < StandardError; end

      def initialize(path)
        @path = path
      end

      def each_instruction
        return enum_for(:each_instruction) unless block_given?

        CSV.foreach(@path, headers: true).with_index(2) do |row, row_number|
          yield parse_row(row, row_number)
        end
      rescue Errno::ENOENT
        raise ParseError, "Transactions file not found: #{@path}"
      rescue CSV::MalformedCSVError => e
        raise ParseError, "Malformed transactions CSV: #{e.message}"
      end

      def read
        each_instruction.to_a
      end

      private

      def parse_row(row, row_number)
        from_value = row['From']
        to_value = row['To']
        amount_value = row['Amount']

        if from_value.nil? || to_value.nil? || amount_value.nil?
          raise ParseError, "Row #{row_number}: missing From, To, or Amount"
        end

        from_account_number = Domain::AccountNumber.new(from_value)
        to_account_number = Domain::AccountNumber.new(to_value)
        amount = Domain::Money.new(amount_value)

        transfer_id = optional_transfer_id(row['TransferId'])

        Domain::TransferInstruction.new(
          from_account_number: from_account_number,
          to_account_number: to_account_number,
          amount: amount,
          transfer_id: transfer_id
        )
      rescue Domain::InvalidAccountNumberError, Domain::InvalidMoneyAmountError,
             Domain::InvalidTransferInstructionError => e
        raise ParseError, "Row #{row_number}: #{e.message}"
      end

      def optional_transfer_id(value)
        return nil if value.nil? || value.strip.empty?

        Domain::TransferId.parse(value)
      end
    end
  end
end

File: nodejs-typescript/src/infrastructure/CsvTransferInstructionReader.ts
import { AccountNumber } from '../domain/AccountNumber.js';
import { Money } from '../domain/Money.js';
import {
  createTransferInstruction,
  type TransferInstruction,
} from '../domain/TransferInstruction.js';
import { TransferId } from '../domain/TransferId.js';
import type { TransferInstructionReader } from '../application/ports/readers.js';
import { CsvParseError, csvParseError } from './errors.js';
import { columnIndex, readCsvRecords, rowNumber } from './csv.js';

export class CsvTransferInstructionReader implements TransferInstructionReader {
  constructor(private readonly path: string) {}

  eachInstruction(callback: (instruction: TransferInstruction) => void): void {
    let records: ReturnType<typeof readCsvRecords>;
    try {
      records = readCsvRecords(this.path);
    } catch (error) {
      throw csvParseError(`Transactions file not found: ${this.path}`, error);
    }

    const fromIndex = columnIndex(records.headers, 'From');
    const toIndex = columnIndex(records.headers, 'To');
    const amountIndex = columnIndex(records.headers, 'Amount');
    const transferIdIndex = records.headers.findIndex(
      (header) => header.trim().toLowerCase() === 'transferid',
    );

    records.rows.forEach((row, index) => {
      callback(
        this.parseRow(row, rowNumber(index), fromIndex, toIndex, amountIndex, transferIdIndex),
      );
    });
  }

  readAll(): TransferInstruction[] {
    const instructions: TransferInstruction[] = [];
    this.eachInstruction((instruction) => {
      instructions.push(instruction);
    });
    return instructions;
  }

  private parseRow(
    row: string[],
    lineNumber: number,
    fromIndex: number,
    toIndex: number,
    amountIndex: number,
    transferIdIndex: number,
  ): TransferInstruction {
    const fromValue = row[fromIndex];
    const toValue = row[toIndex];
    const amountValue = row[amountIndex];
    if (!fromValue || !toValue || !amountValue) {
      throw new CsvParseError(`Row ${String(lineNumber)}: missing From, To, or Amount`);
    }

    try {
      const transferIdValue = transferIdIndex >= 0 ? row[transferIdIndex] : undefined;
      const transferId =
        transferIdValue && transferIdValue.trim().length > 0
          ? TransferId.parse(transferIdValue)
          : undefined;
      return createTransferInstruction({
        fromAccountNumber: AccountNumber.parse(fromValue),
        toAccountNumber: AccountNumber.parse(toValue),
        amount: Money.fromDecimalString(amountValue),
        ...(transferId !== undefined ? { transferId } : {}),
      });
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      throw csvParseError(`Row ${String(lineNumber)}: ${message}`, error);
    }
  }
}
